File: Code/Java/splitarraylargestsum.java
/*
-- Split Array Largest Sum --
Given an array nums which consists of non-negative integers and an integer m,
you can split the array into m non-empty continuous subarrays.
Here is an algorithm to minimize the largest sum among these m subarrays in which
Set left to be the largest number in the array. Set right to be the sum of the array.
Then the result should be in the range of [left, right].
Finding the target by using binary search.
*/

package arrayls;
import java.util.Scanner;

public class splitarr {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        System.out.println("Enter Size of Arraay");
        int n = scan.nextInt();
        int m = scan.nextInt();
        int[] arrr = new int[n];
        System.out.println("Enter "+n+" Elements of array");
        for (int i = 0;i<n;i++){
            arrr[i] = scan.nextInt();
        }
        System.out.println(splitArray(arrr, m));
    }
    public static int splitArray(int[] nums, int m) {
        int n = nums.length;
        int left = 0;
        int right = 0;
        // Here Lest will be incresed Forwards
        //Set left to be the largest number in the array
        //Set right to be the sum of the array.
        for (int i = 0; i < n; ++i) {
            left = Math.max(left, nums[i]);
            right += nums[i];
        }
        // Then the result should be in the range of [left, right].
        while (left <= right) {
            int mid = (right - left) / 2 + left;
            if (canSplit(nums, mid, m)) right = mid - 1;
            else left = mid + 1;
        }
        return left;
    }

    private static boolean canSplit(int[] nums, int target, int m) {
        int cnt = 1;
        int sum = 0;
        for (int i = 0; i < nums.length; ++i) {
            sum += nums[i];
            if (sum > target) {
                ++cnt;
                sum = nums[i];
                if (cnt > m) return false;
            }
        }
        return true;
    }
}
/*
    Test Cases:
        Input : 5   // Length of Array
                2   // value of m
                7 2 5 10 8
        Output: 18

        Input : 3   // Length of Array
                3   // value of m
                1 4 4
        Output: 9

        Time Complexity: O(n) where n is length of array
        Space Complexity: O(1)

 */

File: Code/C++/splitarraylargestsum.cpp
// Given an array nums which consists of non-negative integers and an integer m,
// you can split the array into m non-empty continuous subarrays.
// Input: nums = [7,2,5,10,8], m = 2
// Output: 18
// Explanation:
// There are four ways to split nums into two subarrays.
// The best way is to split it into [7,2,5] and [10,8],
// where the largest sum among the two subarrays is only 18.

#include<bits/stdc++.h>
using namespace std;

int splitArray(vector<int>& nums, int m) {
        int n = nums.size();
        vector<vector<int>> dp(n + 1, vector<int>(m + 1, INT_MAX));
        vector<int> sub(n + 1, 0);
        for (int i = 0; i < n; i++) {
            sub[i + 1] = sub[i] + nums[i];
        }
        // f[i][j] to be the minimum largest subarray sum for splitting nums[0..i] into j parts
        dp[0][0] = 0;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                for (int k = 0; k < i; k++) {
                    dp[i][j] = min(dp[i][j], max(dp[k][j - 1], sub[i] - sub[k]));
                }
            }
        }
        return dp[n][m];
}

int main(){
    // Number of element in the array
    int n;
    cin >> n;
    vector<int> nums(n);
    // Input element of an array
    for(int i = 0; i < n; i++){
        cin >> nums[i];
    }

    int number_of_split;
    cin >> number_of_split;
    // calling splitArray function to find number of ways to split nums into number_of_split subarray
    // with minimum largest sum.
    cout << splitArray(nums, number_of_split) << endl;
    
    return 0;
}

// Input 1: nums = [7,2,5,10,8], number_of_split = 2
// Output 1: 18

// Input 2: nums = [1,2,3,4,5], number_of_split = 2
// Output 2: 9

// Input: nums = [1,4,4],  number_of_split = 3
// Output: 4

// Time Complexity : O(n * n + number_of_split)
// Space Complexity : O(n * number_of_split)
// where n is the size of array nums.
